crates/red_knot_python_semantic/resources/mdtest/metaclass.md

```diff
@@ -0,0 +1,182 @@
+## Default
+
+```py
+class C(type): ...
+
+reveal_type(C.__class__)  # revealed: Literal[type]
+```
+
+## `object`
+
+```py
+reveal_type(object.__class__)  # revealed: Literal[type]
+```
+
+## `type`
+
+```py
+reveal_type(type.__class__)  # revealed: Literal[type]
+```
+
+## Basic
+
+```py
+class A(type): ...
+class B(metaclass=A): ...
+
+reveal_type(B.__class__)  # revealed: Literal[A]
+```
+
+## Invalid metaclass
+
+If a class is a subclass of a class with a custom metaclass, then the subclass will also have that
+metaclass.
+
+```py
+class A: ...
+class B(metaclass=A): ...
+
+# TODO: emit a diagnostic for the invalid metaclass
+reveal_type(B.__class__)  # revealed: Literal[A]
+```
+
+## Linear inheritance
+
+If a class is a subclass of a class with a custom metaclass, then the subclass will also have that
+metaclass.
+
+```py
+class A(type): ...
+class B(metaclass=A): ...
+class C(B): ...
+
+reveal_type(C.__class__)  # revealed: Literal[A]
+```
+
+## Conflict (1)
+
+The metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its
+bases. ("Strict subclass" is a synonym for "proper subclass"; a non-strict subclass can be a
+subclass or the class itself.)
+
+```py
+class A(type): ...
+class B(metaclass=A): ...
+class C(type): ...
+class D(metaclass=C): ...
+
+# error: [conflicting-metaclass] "The metaclass of a derived class (`E`) must be a subclass of the metaclasses of all its bases, but `Literal[C]` and `Literal[A]` have no subclass relationship"
```

Comment:
I think I would prefer for this error message to be

```suggestion
# error: [conflicting-metaclass] "The metaclass of a derived class (`E`) must be a subclass of the metaclasses of all its bases, but `C` and `A` have no subclass relationship"
```

since the error is about a lack of a subclass relationship between classes rather than a subtype relationship between types. (`Literal[A]` is how we would describe the class-literal type with only the class object `A` in it, but `A` is how we would describe the class object `A`.) You could get there with this diff:

```diff
--- a/crates/red_knot_python_semantic/src/types.rs
+++ b/crates/red_knot_python_semantic/src/types.rs
@@ -2123,8 +2123,8 @@ impl<'db> Class<'db> {
                 }
                 return Err(MetaclassError {
                     kind: MetaclassErrorKind::Conflict {
-                        metaclass1: candidate,
-                        metaclass2: metaclass,
+                        metaclass1: candidate.class,
+                        metaclass2: metaclass.class,
                     },
                 });
             }
@@ -2294,8 +2294,8 @@ pub(super) enum MetaclassErrorKind<'db> {
     /// The metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all
     /// its bases.
     Conflict {
-        metaclass1: ClassLiteralType<'db>,
-        metaclass2: ClassLiteralType<'db>,
+        metaclass1: Class<'db>,
+        metaclass2: Class<'db>,
     },
     /// The class inherits from itself!
     ///
diff --git a/crates/red_knot_python_semantic/src/types/infer.rs b/crates/red_knot_python_semantic/src/types/infer.rs
index 0c0991ec4..5287df8d8 100644
--- a/crates/red_knot_python_semantic/src/types/infer.rs
+++ b/crates/red_knot_python_semantic/src/types/infer.rs
@@ -540,8 +540,8 @@ impl<'db> TypeInferenceBuilder<'db> {
                     format_args!(
                         "The metaclass of a derived class (`{}`) must be a subclass of the metaclasses of all its bases, but `{}` and `{}` have no subclass relationship",
                         class.name(self.db),
-                        Type::ClassLiteral(*metaclass1).display(self.db),
-                        Type::ClassLiteral(*metaclass2).display(self.db),
+                        metaclass1.name(self.db),
+                        metaclass2.name(self.db),
                     ),
                 ),
                 MetaclassErrorKind::CyclicDefinition => {
```